Basic Async Unary implementation of QPS benchmark

PeterAdams-A opened the pull request:
Motivation:

The main grpc repo contains a benchmarking suite which can be implemented in any language.
This is an implementation of the async unary test and the framework surrounding it.

Notes:
  * Benchmark so far only run on OSX, not linux
  * Benchmark is not optimised - there are currently allocations within the benchmark which
    could be removed.

Modifications:

Add framework for QPS style testing.
Add implementation of Async Unary benchmark (without TLS)
Add brief README describing how to run the benchmark

Result:

It's possible to run a benchmark of gRPC-swift which is in some way comparible to other
languages.

PeterAdams-A commented:
Note:  All the proto files in the Model directory are taken from grpc/grpc repo - all the swift to match is generated.

glbrntt commented on Performance/QPSBenchmark/Tests/BenchmarkUtilsTests/HistogramTests.swift:
```diff
@@ -0,0 +1,103 @@
+                XCTAssertFalse(twoSeen)
+                twoSeen = true
+            default:
+                XCTAssertTrue(false)
```
You can use `XCTFail` here

glbrntt commented on Performance/QPSBenchmark/Makefile:
```diff
@@ -0,0 +1,74 @@
+# Where products will be built; this is the SPM default.
+GRPC_SWIFT_PATH:=../..
+SWIFT_BUILD_PATH:=./.build
+SWIFT_BUILD_CONFIGURATION=debug
```
let's make this release

glbrntt commented on Performance/QPSBenchmark/Makefile:
```diff
@@ -0,0 +1,74 @@
+PROTOC_GEN_GRPC_SWIFT=${GRPC_SWIFT_PATH}/${SWIFT_BUILD_PATH}/release/protoc-gen-grpc-swift
+
+SWIFT_BUILD:=${SWIFT} build ${SWIFT_FLAGS}
+SWIFT_BUILD_RELEASE:=${SWIFT} build ${SWIFT_FLAGS_RELEASE}
+SWIFT_TEST:=${SWIFT} test ${SWIFT_FLAGS}
+SWIFT_PACKAGE:=${SWIFT} package ${SWIFT_FLAGS}
```
Let's get rid of these too

glbrntt replied:
This one hasn't been addressed.

Although I'm also wondering whether we should just piggyback on top of the main `Makefile` in the root of the repo to generate the protos. WDYT?

glbrntt commented on Performance/QPSBenchmark/Makefile:
```diff
@@ -0,0 +1,74 @@
+SWIFT_BUILD_CONFIGURATION=debug
+SWIFT_FLAGS=--build-path=${SWIFT_BUILD_PATH} --configuration=${SWIFT_BUILD_CONFIGURATION}
+# Force release configuration (for plugins)
+SWIFT_FLAGS_RELEASE=$(patsubst --configuration=%,--configuration=release,$(SWIFT_FLAGS))
```
Let's get rid of this

glbrntt commented on Performance/QPSBenchmark/Makefile:
```diff
@@ -0,0 +1,74 @@
+SWIFT_PACKAGE:=${SWIFT} package ${SWIFT_FLAGS}
+
+# Name of generated xcodeproj
+XCODEPROJ:=GRPC.xcodeproj
```
And this one

glbrntt commented on Performance/QPSBenchmark/Makefile:
```diff
@@ -0,0 +1,74 @@
+
+### Package and plugin build targets ###########################################
+
+all:
+	${SWIFT_BUILD}
+
+.PHONY:
+plugins: ${PROTOC_GEN_SWIFT} ${PROTOC_GEN_GRPC_SWIFT}
+	cp $^ .
```
we can get rid of these too, I think

PeterAdams-A commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/ClientUtilities.swift:
```diff
@@ -0,0 +1,50 @@
+                    return HostAndPort(host: String(host), port: port)
+                }
+            }
+            throw GRPCStatus(code: .invalidArgument, message: "Server targets could not be parsed")
```
We could do.  But then I'd probably want to convert it back to an invalidArgument for sending back to the client as it was an argument which came from the client.  For discussion.

PeterAdams-A commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/AsyncClient.swift:
```diff
@@ -0,0 +1,275 @@
+                                                    config: config,
+                                                    eventLoopGroup: eventLoopGroup))
+        }
+        self.channelRepeaters = channelRepeaters
```
Yes, sadly means I need  a local for eventLoopGroup to avoid self capture but still feels better.

PeterAdams-A commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/BenchmarkServiceImpl.swift:
```diff
@@ -0,0 +1,91 @@
+    private static func makePayload(type: Grpc_Testing_PayloadType, size: Int) throws -> Grpc_Testing_Payload {
+        if type != .compressable {
+            throw GRPCStatus(code: .internalError, message: "Failed to make payload")
+        }
```
I'll comment - I think you have to make data which doesn't compress which is hard (and not implemented in the other implementations I looked at)

PeterAdams-A commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/QPSWorker.swift:
```diff
@@ -0,0 +1,70 @@
+import Logging
+
+/// Sets up and runs a worker service which listens for instructions on what tests to run.
+/// Currently doesn't understand TLS.
```
Probably not worth adding - when I wrote this I was actually looking at the code which does TLS to the driver which we may never want to do.  I'll change this command and add TODOs in client and server code for the other.

PeterAdams-A commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/AsyncServer.swift:
```diff
@@ -0,0 +1,104 @@
+                }
+            }
+        }
+        return promise.futureResult
```
After much confusion I've realised that flatMap change the event loop your future is on.  Now I understand this I'll try and see if I can come up with a better answer.

PeterAdams-A commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/AsyncClient.swift:
```diff
@@ -0,0 +1,275 @@
+        let allStopped = EventLoopFuture<Void>.reduce((),
+                                                      stoppedFutures,
+                                                      on: callbackLoop, { (_, _) -> Void in return () } )
+        _ = allStopped.always { result in
```
I doubt this is safe as I think it will convert the future to another event loop - I'll see if I can come up with something that works.

PeterAdams-A commented:
I've addressed most of your code review issues apart from the few things I've commented on above indicating I will not.  I have added a tiny amount of extra docs as part of breaking up functions.  Is there anything in particular you feel is definitely needing more doc?

glbrntt commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/AsyncServer.swift:
```diff
@@ -0,0 +1,106 @@
+    result.stats.totalCpuTime = 0
+    result.stats.idleCpuTime = 0
+    result.stats.cqPollCount = 0
+    self.logger.info("Sending response")
```
more useful to say "Sending server status" here, "response" is a little ambiguous

glbrntt commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/AsyncServer.swift:
```diff
@@ -0,0 +1,106 @@
+  func shutdown(callbackLoop: EventLoop) -> EventLoopFuture<Void> {
+    return self.server.flatMap { server in
+      server.close()
+    }.hop(to: callbackLoop).flatMap { _ in
```
We should probably stick a `recover` in here too in case the server errors on close.

glbrntt commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/BenchmarkServiceImpl.swift:
```diff
@@ -0,0 +1,111 @@
+  /// Repeated sequence of one request followed by one response.
+  /// Should be called streaming ping-pong
+  /// The server returns the client payload as-is on each response
+  func streamingCall(context: StreamingResponseCallContext<Grpc_Testing_SimpleResponse>)
+    -> EventLoopFuture<(StreamEvent<Grpc_Testing_SimpleRequest>) -> Void> {
```
This is probably the work of swift format, but preferred style in gRPC is:

```swift
func streamingCall(
  context: StreamingResponseCallContext<Grpc_Testing_SimpleResponse>
) -> EventLoopFuture<(StreamEvent<Grpc_Testing_SimpleRequest>) -> Void> {
```

i.e. one argument per line and closing `)` on a new line (since it breaks visually splits signature a body a little better.

I'm fine with these as-is given they're not implemented so just an FYI.

glbrntt commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/BenchmarkServiceImpl.swift:
```diff
@@ -0,0 +1,111 @@
+  private static func makePayload(type: Grpc_Testing_PayloadType,
+                                  size: Int) throws -> Grpc_Testing_Payload {
+    if type != .compressable {
+      // Making a payload which is not compressable is hard - and not implemented in
+      // other implementations too.
```
I guess they planned on doing this but never did and didn't want to break API 🤷‍♂️

glbrntt commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/WorkerServiceImpl.swift:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  /// Create a client of the requested type.
+  private static func createClient(
```
nit: `makeClient`

glbrntt commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/grpcTime.swift:
```diff
@@ -0,0 +1,50 @@
+extension Nanoseconds: CustomStringConvertible {
+  /// Description to aid debugging.
+  var description: String {
+    return String(self.value)
```
Should we include the units too?

```suggestion
    return "\(self.value) ns"
```

glbrntt commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/main.swift:
```diff
@@ -0,0 +1,73 @@
+    logger.info("Worker has finished.")
+  }
+}
+
```
Can we add in the following?

```swift
assert({
  print("⚠️ WARNING: YOU ARE RUNNING IN DEBUG MODE ⚠️")
  return true
}())
```

glbrntt reviewed (changes requested):
Thanks for adding some more docs and splitting things up; the smaller functions help a lot.

This looks pretty good, I've left some comments mostly in `AsyncClient` which I think we can simplify a little.

PeterAdams-A commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/AsyncClient.swift:
```diff
@@ -0,0 +1,293 @@
+      return self.stats.copyData(reset: reset)
+    }
+
+    private static func makeClientRequest(payloadConfig: Grpc_Testing_PayloadConfig) throws
```
I think having looked at it, that this would make us more similar to the C version too.

glbrntt commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/AsyncClient.swift:
```diff
@@ -192,7 +238,7 @@ final class AsyncUnaryQPSClient: QPSClient {
       } else {
         // Try scheduling another request.
-        try! self.launchRequests()
+        try! self.makeRequestAndRepeat()
```
This function only `throws` because of this `try` on itself, let's drop both!

glbrntt commented on Performance/QPSBenchmark/Sources/QPSBenchmark/Runtime/AsyncClient.swift:
```diff
@@ -150,23 +191,28 @@ final class AsyncUnaryQPSClient: QPSClient {
     }
 
+    /// Returns if it is permissible to make another request - ie we've not been asked to stop, and we're not at the limit of outstanding requests.
+    private func canMakeRequest() -> Bool {
```
This should be a computed property: 

```swift
private var canMakeRequest: Bool {
  return !(self.stopRequested || self.numberOfOutstandingRequests >= self.maxPermittedOutstandingRequests)
}
```

glbrntt reviewed (changes requested):
Looks great, just two nits remaining

glbrntt reviewed (approved):
LGTM thanks @PeterAdams-A!
